File: frontend/scripts/coverage-provenance.mjs
// Run the real frontend suite and bind its report to unchanged source bytes.
import { createHash } from 'node:crypto';
import { lstatSync, readdirSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { spawnSync } from 'node:child_process';

const digest = (bytes) => createHash('sha256').update(bytes).digest('hex');
function snapshot(path = 'src', files = {}) {
  const stat = lstatSync(path);
  if (stat.isSymbolicLink()) throw new Error('Symlink in coverage source');
  if (stat.isDirectory()) {
    for (const name of readdirSync(path).sort()) snapshot(`${path}/${name}`, files);
  } else if (stat.isFile() && /\.(ts|tsx)$/.test(path)) {
    files[path] = digest(readFileSync(path));
  }
  return files;
}

rmSync('coverage/provenance.json', { force: true });
rmSync('coverage/coverage-summary.json', { force: true });
const before = snapshot();
if (!Object.keys(before).length) throw new Error('Empty coverage source');
const result = spawnSync('npm', ['test', ...process.argv.slice(2)], { stdio: 'inherit' });
if (result.error || result.status !== 0) process.exit(result.status || 1);
if (JSON.stringify(before) !== JSON.stringify(snapshot())) {
  throw new Error('Source changed during coverage run');
}
writeFileSync('coverage/provenance.json', JSON.stringify({
  version: 1,
  scope: 'frontend',
  sources: before,
  report_sha256: digest(readFileSync('coverage/coverage-summary.json')),
}));

File: scripts/coverage_provenance.py
"""Bind a successful coverage run to unchanged application source and report bytes."""

import hashlib
import json
import subprocess
import sys
from pathlib import Path


def report_path(scope, root):
    names = {"backend": "coverage.json", "frontend": "coverage-summary.json"}
    if scope not in names:
        raise ValueError("Unknown scope")
    return Path(root) / "coverage" / names[scope]


def snapshot(scope, root):
    report_path(scope, root)  # Validate scope before choosing source paths.
    root = Path(root)
    source = root / ("app" if scope == "backend" else "src")
    suffixes = {".py"} if scope == "backend" else {".ts", ".tsx"}
    paths = [source, *source.rglob("*")]
    if any(path.is_symlink() for path in paths):
        raise ValueError("Symlink in coverage source")
    files = {
        path.relative_to(root).as_posix(): hashlib.sha256(path.read_bytes()).hexdigest()
        for path in paths
        if path.is_file() and path.suffix in suffixes
    }
    if not files:
        raise ValueError("Empty coverage source")
    return files


def record(scope, root, before):
    if snapshot(scope, root) != before:
        raise ValueError("Source changed during coverage run")
    report = report_path(scope, root)
    payload = {
        "version": 1,
        "scope": scope,
        "sources": before,
        "report_sha256": hashlib.sha256(report.read_bytes()).hexdigest(),
    }
    report.with_name("provenance.json").write_text(json.dumps(payload, sort_keys=True))


def verify(scope, root):
    report = report_path(scope, root)
    # Read once: the checker parses these exact digest-verified bytes.
    data = report.read_bytes()
    payload = json.loads(report.with_name("provenance.json").read_text())
    expected = {
        "version": 1,
        "scope": scope,
        "sources": snapshot(scope, root),
        "report_sha256": hashlib.sha256(data).hexdigest(),
    }
    if (
        not isinstance(payload, dict)
        or type(payload.get("version")) is not int
        or payload != expected
    ):
        raise ValueError("Stale or invalid coverage provenance")
    return data


if __name__ == "__main__":
    scope, directory, *command = sys.argv[1:]
    report_path(scope, directory).with_name("provenance.json").unlink(missing_ok=True)
    report_path(scope, directory).unlink(missing_ok=True)
    before = snapshot(scope, directory)
    subprocess.run(command, cwd=directory, check=True)
    record(scope, directory, before)
